File: 01-Programming-Languages/Scala/OOP/classes.scala
/**
 * Shape Hierarchy in Scala
 *
 * Demonstrates Scala OOP: traits, case classes, abstract classes,
 * companion objects, pattern matching, and mixins.
 *
 * Run: scala classes.scala
 */

// ============================================================
// Traits: Interfaces + implementation
// ============================================================
trait Drawable {
  def draw(): String
}

trait Resizable {
  def resize(factor: Double): Unit
}

trait PrettyPrint {
  def prettyPrint: String
}

// ============================================================
// Abstract class: Common shape behavior
// ============================================================
abstract class Shape(val color: String = "black") extends Drawable with PrettyPrint {
  def area: Double
  def perimeter: Double
  def typeName: String

  def describe: String =
    f"$typeName [$color]: area=$area%.2f, perimeter=$perimeter%.2f"

  override def prettyPrint: String = describe

  // Compare shapes by area
  def compareArea(other: Shape): Int = this.area.compareTo(other.area)
}

// ============================================================
// Case classes: Immutable data carriers
// ============================================================
case class Circle(
  radius: Double,
  color: String = "black"
) extends Shape(color) with Resizable {

  override def area: Double = Math.PI * radius * radius
  override def perimeter: Double = 2 * Math.PI * radius
  override def typeName: String = "Circle"

  override def draw(): String =
    s"Drawing circle r=$radius [color=$color]"

  override def resize(factor: Double): Unit =
    // Case classes are immutable, so this creates a new copy
    // For mutability, use a regular class with var
    println(s"Cannot resize immutable case class. Use copy() instead.")

  def resized(factor: Double): Circle = copy(radius = radius * factor)
}

case class Rectangle(
  width: Double,
  height: Double,
  color: String = "black"
) extends Shape(color) with Resizable {

  override def area: Double = width * height
  override def perimeter: Double = 2 * (width + height)
  override def typeName: String =
    if (width == height) "Square" else "Rectangle"

  override def draw(): String =
    s"Drawing rectangle ${width}x$height [color=$color]"

  override def resize(factor: Double): Unit =
    println(s"Cannot resize immutable case class. Use copy() instead.")

  def resized(factor: Double): Rectangle = copy(width = width * factor, height = height * factor)
}

case class Triangle(
  a: Double,
  b: Double,
  c: Double,
  color: String = "black"
) extends Shape(color) {

  override def area: Double = {
    val s = (a + b + c) / 2
    Math.sqrt(s * (s - a) * (s - b) * (s - c))
  }

  override def perimeter: Double = a + b + c
  override def typeName: String = "Triangle"

  override def draw(): String =
    s"Drawing triangle ($a, $b, $c) [color=$color]"
}

// ============================================================
// Companion object: Factory methods and constants
// ============================================================
object Shape {
  // Factory methods
  def circle(radius: Double, color: String = "black"): Circle =
    Circle(radius, color)

  def rectangle(width: Double, height: Double, color: String = "black"): Rectangle =
    Rectangle(width, height, color)

  def square(side: Double, color: String = "black"): Rectangle =
    Rectangle(side, side, color)

  def triangle(a: Double, b: Double, c: Double, color: String = "black"): Triangle =
    Triangle(a, b, c, color)

  // Validate triangle sides
  def isValidTriangle(a: Double, b: Double, c: Double): Boolean =
    a + b > c && a + c > b && b + c > a

  // Constants
  val PI: Double = Math.PI
}

// ============================================================
// Shape collection with functional operations
// ============================================================
class ShapeCollection(shapes: List[Shape] = List.empty) {
  def add(shape: Shape): ShapeCollection =
    new ShapeCollection(shapes :+ shape)

  def sortedByArea: List[Shape] = shapes.sortBy(_.area)

  def totalArea: Double = shapes.map(_.area).sum

  def filterByType[T <: Shape](implicit manifest: scala.reflect.ClassTag[T]): List[Shape] =
    shapes.filter(_.isInstanceOf[T])

  def largest: Option[Shape] = shapes.maxByOption(_.area)
  def smallest: Option[Shape] = shapes.minByOption(_.area)

  def count: Int = shapes.size
  def all: List[Shape] = shapes
}

// ============================================================
// Pattern matching with sealed traits
// ============================================================
sealed trait ShapeCategory
case object SmallShape extends ShapeCategory   // area < 10
case object MediumShape extends ShapeCategory  // 10 <= area < 50
case object LargeShape extends ShapeCategory   // area >= 50

def categorize(shape: Shape): ShapeCategory =
  shape.area match {
    case a if a < 10  => SmallShape
    case a if a < 50  => MediumShape
    case _            => LargeShape
  }

// ============================================================
// Demo
// ============================================================
object ShapeDemo extends App {
  println("=== Scala Shape Hierarchy ===\n")

  // Create shapes using companion object
  val shapes = List(
    Shape.circle(5, "red"),
    Shape.rectangle(4, 6, "blue"),
    Shape.square(3, "green"),
    Shape.triangle(3, 4, 5, "orange")
  )

  // Display all shapes
  println("All Shapes:")
  println("-" * 60)
  shapes.foreach(s => println(s"  ${s.describe}"))

  // Pattern matching
  println("\nDrawing shapes:")
  shapes.foreach { shape =>
    shape match {
      case Circle(r, c) if r > 4  => println(s"  Large circle! ${shape.draw()}")
      case Circle(r, c)           => println(s"  ${shape.draw()}")
      case Rectangle(w, h, c)     => println(s"  ${shape.draw()}")
      case t: Triangle            => println(s"  ${shape.draw()}")
    }
  }

  // Functional operations
  println("\nSorted by area (ascending):")
  shapes.sortBy(_.area).foreach(s =>
    println(f"  ${s.typeName}: ${s.area}%.2f")
  )

  // Collection
  val collection = shapes.foldLeft(new ShapeCollection())(_.add(_))
  println(s"\nTotal shapes: ${collection.count}")
  println(f"Total area: ${collection.totalArea}%.2f")

  // Categorize
  println("\nShape categories:")
  shapes.foreach { shape =>
    val category = categorize(shape)
    println(s"  ${shape.typeName}: $category")
  }

  // Resizing (creates new copy for case classes)
  println("\nResizing:")
  val original = Shape.circle(5)
  val resized = original.resized(2)
  println(f"  Original: r=${original.radius}%,.2f, area=${original.area}%.2f")
  println(f"  Resized:  r=${resized.radius}%,.2f, area=${resized.area}%.2f")

  // Pattern matching with copy
  println("\nCase class pattern matching:")
  val Circle(r, c) = Shape.circle(10, "red")
  println(s"  Extracted: radius=$r, color=$c")
}

File: 01-Programming-Languages/PHP/OOP/classes.php
<?php
/**
 * Shape Hierarchy in PHP
 *
 * Demonstrates PHP OOP features: abstract classes, interfaces,
 * traits, and concrete implementations.
 *
 * Run: php classes.php
 */

// ============================================================
// Interface: Defines contracts for shapes
// ============================================================
interface Drawable {
    public function draw(): string;
}

interface Resizable {
    public function resize(float $factor): void;
}

interface Comparable {
    public function compareTo(self $other): int;
}

// ============================================================
// Traits: Reusable behavior
// ============================================================
trait Timestampable {
    private string $createdAt;

    public function __constructTimestampable(): void {
        $this->createdAt = date('Y-m-d H:i:s');
    }

    public function getCreatedAt(): string {
        return $this->createdAt;
    }
}

trait Printable {
    public function printDescription(): void {
        echo $this->describe() . "\n";
    }
}

// ============================================================
// Abstract class: Common shape behavior
// ============================================================
abstract class Shape implements Drawable, Comparable {
    use Timestampable, Printable;

    protected float $x;
    protected float $y;
    protected string $color;

    public function __construct(float $x = 0, float $y = 0, string $color = "black") {
        $this->x = $x;
        $this->y = $y;
        $this->color = $color;
        $this->__constructTimestampable();
    }

    abstract public function area(): float;
    abstract public function perimeter(): float;
    abstract public function type(): string;

    public function position(): string {
        return "({$this->x}, {$this->y})";
    }

    public function describe(): string {
        return sprintf(
            "%s [%s] at %s: area=%.2f, perimeter=%.2f",
            $this->type(),
            $this->color,
            $this->position(),
            $this->area(),
            $this->perimeter()
        );
    }

    public function draw(): string {
        return "Drawing {$this->type()} at {$this->position()}";
    }

    public function compareTo(Shape $other): int {
        return $this->area() <=> $other->area();
    }
}

// ============================================================
// Concrete shapes
// ============================================================
class Circle extends Shape implements Resizable {
    private float $radius;

    public function __construct(
        float $radius,
        float $x = 0,
        float $y = 0,
        string $color = "black"
    ) {
        parent::__construct($x, $y, $color);
        $this->radius = $radius;
    }

    public function area(): float {
        return M_PI * $this->radius ** 2;
    }

    public function perimeter(): float {
        return 2 * M_PI * $this->radius;
    }

    public function type(): string {
        return "Circle";
    }

    public function radius(): float {
        return $this->radius;
    }

    public function resize(float $factor): void {
        $this->radius *= $factor;
    }

    public function draw(): string {
        return "Drawing circle r={$this->radius} at {$this->position()}";
    }
}

class Rectangle extends Shape implements Resizable {
    private float $width;
    private float $height;

    public function __construct(
        float $width,
        float $height,
        float $x = 0,
        float $y = 0,
        string $color = "black"
    ) {
        parent::__construct($x, $y, $color);
        $this->width = $width;
        $this->height = $height;
    }

    public function area(): float {
        return $this->width * $this->height;
    }

    public function perimeter(): float {
        return 2 * ($this->width + $this->height);
    }

    public function type(): string {
        return "Rectangle";
    }

    public function width(): float { return $this->width; }
    public function height(): float { return $this->height; }

    public function resize(float $factor): void {
        $this->width *= $factor;
        $this->height *= $factor;
    }
}

class Square extends Rectangle {
    public function __construct(
        float $side,
        float $x = 0,
        float $y = 0,
        string $color = "black"
    ) {
        parent::__construct($side, $side, $x, $y, $color);
    }

    public function type(): string {
        return "Square";
    }

    public function side(): float {
        return $this->width();
    }
}

class Triangle extends Shape {
    private float $a;
    private float $b;
    private float $c;

    public function __construct(
        float $a,
        float $b,
        float $c,
        float $x = 0,
        float $y = 0,
        string $color = "black"
    ) {
        parent::__construct($x, $y, $color);
        $this->a = $a;
        $this->b = $b;
        $this->c = $c;
    }

    public function area(): float {
        $s = ($this->a + $this->b + $this->c) / 2;
        return sqrt($s * ($s - $this->a) * ($s - $this->b) * ($s - $this->c));
    }

    public function perimeter(): float {
        return $this->a + $this->b + $this->c;
    }

    public function type(): string {
        return "Triangle";
    }
}

// ============================================================
// Shape collection using generics-like pattern
// ============================================================
class ShapeCollection {
    private array $shapes = [];

    public function add(Shape $shape): void {
        $this->shapes[] = $shape;
    }

    public function sortByArea(): array {
        $sorted = $this->shapes;
        usort($sorted, fn(Shape $a, Shape $b) => $a->compareTo($b));
        return $sorted;
    }

    public function totalArea(): float {
        return array_reduce($this->shapes, fn(float $sum, Shape $s) => $sum + $s->area(), 0.0);
    }

    public function all(): array {
        return $this->shapes;
    }
}

// ============================================================
// Demo
// ============================================================
echo "=== PHP Shape Hierarchy ===\n\n";

// Create shapes
$shapes = [
    new Circle(5, 0, 0, "red"),
    new Rectangle(4, 6, 1, 1, "blue"),
    new Square(3, 2, 2, "green"),
    new Triangle(3, 4, 5, 3, 3, "orange"),
];

// Display all shapes
echo "All Shapes:\n";
echo str_repeat("-", 60) . "\n";
foreach ($shapes as $shape) {
    $shape->printDescription();
}

// Demonstrate interfaces
echo "\nDrawing shapes:\n";
foreach ($shapes as $shape) {
    echo "  " . $shape->draw() . "\n";
}

// Demonstrate traits
echo "\nCreation times:\n";
foreach ($shapes as $shape) {
    echo "  {$shape->type()}: created at {$shape->getCreatedAt()}\n";
}

// Demonstrate sorting by area
echo "\nSorted by area (ascending):\n";
$collection = new ShapeCollection();
foreach ($shapes as $shape) {
    $collection->add($shape);
}
$sorted = $collection->sortByArea();
foreach ($sorted as $shape) {
    echo "  {$shape->type()}: " . round($shape->area(), 2) . "\n";
}

echo "\nTotal area: " . round($collection->totalArea(), 2) . "\n";

// Demonstrate resizing (Resizable interface)
echo "\nAfter resizing Circle by 2x:\n";
$circle = new Circle(5);
echo "  Before: radius={$circle->radius()}, area=" . round($circle->area(), 2) . "\n";
$circle->resize(2);
echo "  After:  radius={$circle->radius()}, area=" . round($circle->area(), 2) . "\n";

// Demonstrate polymorphism
echo "\nPolymorphism demo:\n";
function printShapeInfo(Shape $shape): void {
    echo "  " . $shape->describe() . "\n";
}

$shapes = [
    new Circle(3),
    new Rectangle(5, 4),
    new Square(6),
];
foreach ($shapes as $shape) {
    printShapeInfo($shape);
}
